File: software/examples/rust/example_callback.rs
use std::{error::Error, io, thread};
use tinkerforge::{ip_connection::IpConnection, outdoor_weather_bricklet::*};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Outdoor Weather Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let ow = OutdoorWeatherBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Enable station data callbacks
    ow.set_station_callback_configuration(true);

    // Enable sensor data callbacks
    ow.set_sensor_callback_configuration(true);

    let station_data_receiver = ow.get_station_data_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `ow` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for station_data in station_data_receiver {
            println!("Identifier (Station): {}", station_data.identifier);
            println!(
                "Temperature (Station): {} °C",
                station_data.temperature as f32 / 10.0
            );
            println!("Humidity (Station): {} %RH", station_data.humidity);
            println!(
                "Wind Speed (Station): {} m/s",
                station_data.wind_speed as f32 / 10.0
            );
            println!(
                "Gust Speed (Station): {} m/s",
                station_data.gust_speed as f32 / 10.0
            );
            println!("Rain (Station): {} mm", station_data.rain as f32 / 10.0);

            if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_N {
                println!("Wind Direction (Station): N");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_NNE {
                println!("Wind Direction (Station): NNE");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_NE {
                println!("Wind Direction (Station): NE");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_ENE {
                println!("Wind Direction (Station): ENE");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_E {
                println!("Wind Direction (Station): E");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_ESE {
                println!("Wind Direction (Station): ESE");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_SE {
                println!("Wind Direction (Station): SE");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_SSE {
                println!("Wind Direction (Station): SSE");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_S {
                println!("Wind Direction (Station): S");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_SSW {
                println!("Wind Direction (Station): SSW");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_SW {
                println!("Wind Direction (Station): SW");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_WSW {
                println!("Wind Direction (Station): WSW");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_W {
                println!("Wind Direction (Station): W");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_WNW {
                println!("Wind Direction (Station): WNW");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_NW {
                println!("Wind Direction (Station): NW");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_NNW {
                println!("Wind Direction (Station): NNW");
            } else if station_data.wind_direction == OUTDOOR_WEATHER_BRICKLET_WIND_DIRECTION_ERROR {
                println!("Wind Direction (Station): Error");
            }

            println!("Battery Low (Station): {}", station_data.battery_low);
            println!();
        }
    });

    let sensor_data_receiver = ow.get_sensor_data_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `ow` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for sensor_data in sensor_data_receiver {
            println!("Identifier (Sensor): {}", sensor_data.identifier);
            println!(
                "Temperature (Sensor): {} °C",
                sensor_data.temperature as f32 / 10.0
            );
            println!("Humidity (Sensor): {} %RH", sensor_data.humidity);
            println!();
        }
    });

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_callback.go
package main

import (
	"fmt"
	"github.com/Tinkerforge/go-api-bindings/ipconnection"
	"github.com/Tinkerforge/go-api-bindings/outdoor_weather_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Outdoor Weather Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	ow, _ := outdoor_weather_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Enable station data callbacks
	ow.SetStationCallbackConfiguration(true)

	// Enable sensor data callbacks
	ow.SetSensorCallbackConfiguration(true)

	ow.RegisterStationDataCallback(func(identifier uint8, temperature int16, humidity uint8, windSpeed uint32, gustSpeed uint32, rain uint32, windDirection outdoor_weather_bricklet.WindDirection, batteryLow bool) {
		fmt.Printf("Identifier (Station): %d\n", identifier)
		fmt.Printf("Temperature (Station): %f °C\n", float64(temperature)/10.0)
		fmt.Printf("Humidity (Station): %d %RH\n", humidity)
		fmt.Printf("Wind Speed (Station): %f m/s\n", float64(windSpeed)/10.0)
		fmt.Printf("Gust Speed (Station): %f m/s\n", float64(gustSpeed)/10.0)
		fmt.Printf("Rain (Station): %f mm\n", float64(rain)/10.0)

		if windDirection == outdoor_weather_bricklet.WindDirectionN {
			fmt.Println("Wind Direction (Station): N")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionNNE {
			fmt.Println("Wind Direction (Station): NNE")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionNE {
			fmt.Println("Wind Direction (Station): NE")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionENE {
			fmt.Println("Wind Direction (Station): ENE")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionE {
			fmt.Println("Wind Direction (Station): E")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionESE {
			fmt.Println("Wind Direction (Station): ESE")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionSE {
			fmt.Println("Wind Direction (Station): SE")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionSSE {
			fmt.Println("Wind Direction (Station): SSE")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionS {
			fmt.Println("Wind Direction (Station): S")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionSSW {
			fmt.Println("Wind Direction (Station): SSW")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionSW {
			fmt.Println("Wind Direction (Station): SW")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionWSW {
			fmt.Println("Wind Direction (Station): WSW")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionW {
			fmt.Println("Wind Direction (Station): W")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionWNW {
			fmt.Println("Wind Direction (Station): WNW")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionNW {
			fmt.Println("Wind Direction (Station): NW")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionNNW {
			fmt.Println("Wind Direction (Station): NNW")
		} else if windDirection == outdoor_weather_bricklet.WindDirectionError {
			fmt.Println("Wind Direction (Station): Error")
		}

		fmt.Printf("Battery Low (Station): %s\n", batteryLow)
		fmt.Println()
	})

	ow.RegisterSensorDataCallback(func(identifier uint8, temperature int16, humidity uint8) {
		fmt.Printf("Identifier (Sensor): %d\n", identifier)
		fmt.Printf("Temperature (Sensor): %f °C\n", float64(temperature)/10.0)
		fmt.Printf("Humidity (Sensor): %d %RH\n", humidity)
		fmt.Println()
	})

	fmt.Print("Press enter to exit.")
	fmt.Scanln()
}
